go-controller/pkg/ovn/libovsdbops/acl.go

```diff
@@ -0,0 +1,220 @@
+	}
+	return ""
+}
+
+// isEquivalentACL if it hase same uuid, or if it has same name
+// and external ids, or if it has same priority, direction, match
+// and action.
+func isEquivalentACL(existing *nbdb.ACL, searched *nbdb.ACL) bool {
+	if searched.UUID != "" && existing.UUID == searched.UUID {
+		return true
+	}
+
+	eName := getACLName(existing)
+	sName := getACLName(searched)
+	// TODO if we want to support adding/removing external ids,
+	// we need to compare them differently, perhaps just the common subset
+	if eName != "" && eName == sName && reflect.DeepEqual(existing.ExternalIDs, searched.ExternalIDs) {
+		return true
+	}
+
+	return existing.Priority == searched.Priority &&
+		existing.Direction == searched.Direction &&
+		existing.Match == searched.Match &&
+		existing.Action == searched.Action
+}
+
+// findACL looks up the client cache for equivalent ACLs
+func findACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) (string, error) {
+	if acl.UUID != "" && !isNamedUUID(acl.UUID) {
+		return acl.UUID, nil
+	}
+
+	acls := []nbdb.ACL{}
+	err := nbClient.WhereCache(func(item *nbdb.ACL) bool {
+		return isEquivalentACL(item, acl)
+	}).List(&acls)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return "", fmt.Errorf("can't find ACL by equivalence %+v: %v", *acl, err)
+	}
+
+	if len(acls) > 1 {
+		return "", fmt.Errorf("unexpectedly found multiple equivalent ACLs: %+v", acls)
+	}
+
+	if len(acls) == 1 {
+		return acls[0].UUID, nil
+	}
+
+	return "", nil
+}
+
+func BuildACL(name, direction string, priority int, match, action, meter, severity string, log bool, externalIds map[string]string) *nbdb.ACL {
+	uuid := buildNamedUUID()
+	name = fmt.Sprintf("%.63s", name)
+	return &nbdb.ACL{
+		UUID:        uuid,
+		Name:        &name,
+		Direction:   direction,
+		Match:       match,
+		Action:      action,
+		Priority:    priority,
+		Severity:    &severity,
+		Log:         log,
+		Meter:       &meter,
+		ExternalIDs: externalIds,
+	}
+}
+
+func CreateOrUpdateACLOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	uuid, err := findACL(nbClient, acl)
+	if err != nil {
+		return nil, err
+	}
+
+	// If ACL already exists, update it
+	if uuid != "" {
+		acl.UUID = ""
+		op, err := nbClient.Where(&nbdb.ACL{UUID: uuid}).Update(acl)
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, op...)
+		acl.UUID = uuid
+		return ops, nil
+	}
+
+	op, err := nbClient.Create(acl)
+	if err != nil {
+		return nil, err
+	}
+	ops = append(ops, op...)
+
+	return ops, nil
+}
+
+func CreateOrUpdateACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) error {
+	ops, err := CreateOrUpdateACLOps(nbClient, nil, acl)
+	if err != nil {
+		return err
+	}
+
+	_, err = TransactAndCheck(nbClient, ops)
+	return err
+}
+
+func UpdateACLOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
```

Comment:
Not really. `UpdateACL` would fail if the ACL to be updated does not exist but indeed there is no use for that actually so yeah, it can be removed.
`UpdateACLLoggingOps` is also different: it only updates severity and log and not the full ACL; and does not require to know the full details of the ACL to be updated. Circumstantially we could have used `CreateOrUpdateACL` in the current cases that we have for this but is not logically what we want to do so I decided to stick with a separate method.
The idea that I had for `libovsdbops`is to provide all the convinience methods that we need vs being a compact api.